pkg/detector/detector.go

```diff
@@ -1117,20 +1125,40 @@ func (d *ResourceDetector) ReconcileClusterPropagationPolicy(key util.QueueKey)
 	unstructuredObj, err := d.clusterPropagationPolicyLister.Get(ckey.NamespaceKey())
 	if err != nil {
 		if apierrors.IsNotFound(err) {
-			klog.Infof("ClusterPropagationPolicy(%s) has been removed.", ckey.NamespaceKey())
-			return d.HandleClusterPropagationPolicyDeletion(ckey.Name)
+			return nil
 		}
 
 		klog.Errorf("Failed to get ClusterPropagationPolicy(%s): %v", ckey.NamespaceKey(), err)
 		return err
 	}
 
-	klog.Infof("Policy(%s) has been added", ckey.NamespaceKey())
 	propagationObject := &policyv1alpha1.ClusterPropagationPolicy{}
 	if err = helper.ConvertToTypedObject(unstructuredObj, propagationObject); err != nil {
 		klog.Errorf("Failed to convert ClusterPropagationPolicy(%s) from unstructured object: %v", ckey.NamespaceKey(), err)
 		return err
 	}
+
+	if !propagationObject.DeletionTimestamp.IsZero() {
+		klog.Infof("ClusterPropagationPolicy(%s) has been deleted.", ckey.NamespaceKey())
+		if err = d.HandleClusterPropagationPolicyDeletion(propagationObject.Name); err != nil {
+			return err
+		}
+		if controllerutil.RemoveFinalizer(propagationObject, util.ClusterPropagationPolicyControllerFinalizer) {
+			if err = d.Client.Update(context.TODO(), propagationObject); err != nil {
+				klog.Errorf("Failed to remove finalizer for ClusterPropagationPolicy(%s), err: %v", ckey.NamespaceKey(), err)
+				return err
+			}
+		}
+		return nil
+	}
+
+	klog.Infof("ClusterPropagationPolicy(%s) has been added.", ckey.NamespaceKey())
+	if controllerutil.AddFinalizer(propagationObject, util.ClusterPropagationPolicyControllerFinalizer) {
+		if err = d.Client.Update(context.TODO(), propagationObject); err != nil {
+			klog.Errorf("Failed to add finalizer for ClusterPropagationPolicy(%s), err: %v", ckey.NamespaceKey(), err)
+			return err
+		}
```

Comment:
The update will trigger the PP update again, does this mean that `HandleClusterPropagationPolicyCreationOrUpdate` will be executed twice .

Reply:
nice catch!
How about just return after the update, that means in the first reconcile loop, we focus on adding the finalizer and sync by the second reconcile loop? 
like:
```diff
diff --git a/pkg/detector/detector.go b/pkg/detector/detector.go
index 893f9f611..5bbc2b2fc 100644
--- a/pkg/detector/detector.go
+++ b/pkg/detector/detector.go
@@ -1158,6 +1158,7 @@ func (d *ResourceDetector) ReconcileClusterPropagationPolicy(key util.QueueKey)
                        klog.Errorf("Failed to add finalizer for ClusterPropagationPolicy(%s), err: %v", ckey.NamespaceKey(), err)
                        return err
                }
+               return nil
        }
        return d.HandleClusterPropagationPolicyCreationOrUpdate(propagationObject)
 }
```

Reply:
+1